### ISSUE TITLE
Upgrading from 2.100 to 2.102 results in server hanging

### ISSUE THREAD

**Issue by nc2y** [kohler/hotcrp#163]
I doubt this is a HotCRP problem _per se_ rather than a poor integration with my system, but I am hoping somebody can help me figure this out before I pull whatever hair I have left. 

I simply `git pull`'ed my HotCRP installation directory. I have multi-conference set up. The previous version (2.100) worked like a charm, but 2.102 results in the server hanging forever on all pages, without showing any data, irrespective of the conference I'm trying to access. So I suspect this is something really trivial the server isn't happy about. 

I confirmed this doesn't look like a PHP issue -- other PHP pages on the same site work just fine.

I mistakenly had `sudo git pull`'ed so all files in the hotcrp directory had been changed to `root.root` ownership. Worried I had mucked the permissions, I reverted ownership to what it should be in my case (`www-data.www-data`) seemingly to no effect.

Any idea on how to even start troubleshooting this? I tried restarting `apache`, `mysqld`, and even rebooted the machine out of frustration. 

Apache/2.4.29 (Ubuntu) - stock Ubuntu 18.04 LTS with PHP 7.2.

**Comment by nc2y** [comment]
Thanks for the quick reply! The error messages and logs I got seem to support this:

1. From Apache2's`error.log`: 
```[Fri Jan 25 17:28:39.070460 2019] [php7:notice] [pid 11293] [client 128.237.152.167:65383] PHP Notice:  A non well formed numeric value encountered in /srv/hotcrp/src/helpers.php on line 705, referer: https://arima.cylab.cmu.edu/bitcoin17/``` 
(a bunch of similar errors...)
```[Fri Jan 25 17:30:21.978559 2019] [php7:notice] [pid 17204] [client 128.237.152.167:65409] bitcoin17: updating schema from version 170, referer: https://arima.cylab.cmu.edu/bitcoin17/?post=9krr0ag0
[Fri Jan 25 17:30:23.655017 2019] [php7:notice] [pid 17204] [client 128.237.152.167:65409] /srv/hotcrp/src/updateschema.php:1309:updateSchema: database warning: Note (1051) Unknown table 'bitcoin17.DeletedContactInfo', referer: https://arima.cylab.cmu.edu/bitcoin17/?post=9krr0ag0
[Fri Jan 25 17:41:40.774503 2019] [php7:notice] [pid 22101] [client 128.237.152.167:49199] bitcoin17: updating schema from version 177
[Fri Jan 25 17:41:40.849255 2019] [php7:notice] [pid 22097] [client 128.237.152.167:65474] bitcoin17: updating schema from version 177```
2 and 3. ```lib/runsql.sh -n bitcoin17
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 39
Server version: 5.7.25-0ubuntu0.18.04.2 (Ubuntu)

Copyright (c) 2000, 2019, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql> select * from Settings where name='__schema_lock';
+---------------+-------+------+
| name          | value | data |
+---------------+-------+------+
| __schema_lock |     1 | NULL |
+---------------+-------+------+
1 row in set (0.00 sec)

mysql> select * from Settings where name='allowPaperOption';
+------------------+-------+------+
| name             | value | data |
+------------------+-------+------+
| allowPaperOption |   177 | NULL |
+------------------+-------+------+
1 row in set (0.00 sec)```

I have a couple of other conferences on the server, which seem to be equally stuck.

**Comment by kohler** [comment]
OK. This sucks, but I think I may understand what's going on.

In HotCRP.com deployments, databases do not actually contain PDF data. The PDFs are stored elsewhere. As a result, commands like `alter table PaperStorage` are relatively quick to complete, because the `PaperStorage` table is small.

But for some users `PaperStorage` is huge. Maybe you are one such user. And for you, a command like `` alter table PaperStorage drop `mimetypeid` `` might take a long time. Long enough for the PHP web thread upgrading the database to be aborted. And then the system is stuck halfway through a schema update.

I'm not sure how the schema updater could/should handle this, but here is how I think you can recover manually.

1. Figure out whether the `alter table PaperStorage` completed. Do a `describe PaperStorage;` and see if it has a `mimetypeid` column.
2. If it *does* have such a column, then skip to step 4.
3. If it *does not* have such a column, then these commands will complete the aborted update: `` drop table if exists `Mimetype`; update Settings set value=178 where name='allowPaperOption' and value=177; ``
4. Remove the schema lock: `delete from Settings where name='__schema_lock'`
5. Run a command-line utility to complete the schema update. This will hopefully avoid getting stuck again, because php command-line utilities don't time out. Example: `php batch/checkinvariants.php -n CONFERENCEID`

I also just checked in 878fcf4759fea5ba7593f3fb9a66748f6870857c, which attempts to code around this specific problem, and also ups the runtime limit higher. If you feel comfortable applying that commit, you could apply it and then just run steps 4 and 5.

Please let me know whether this works.

**Comment by kohler** [comment]
The “non-numeric” warning I think has been fixed since the release of 2.102—if you tell me what commit you're on I can guide further.

**Comment by nc2y** [comment]
It worked! (I didn't use the new commit, I used the step-by-step recipe you gave above.)

Indeed the `php batch/checkinvariants.php -n CONFERENCEID` took a good 3-4 minutes to complete in my case, so your hypothesis was confirmed. It's not that huge a conference (~38 PDFs of what I expect - without checking - are mostly reasonable sizes) but it is hosted on a bit of a wimpy machine...

To be honest, I am not sure it was a server timeout, I might just not have had enough patience when connecting the first time and might have shot myself in the foot by aborting the web connection, somehow causing some weird state on the server side. It may be worth mentioning in the install instructions that running HotCRP for the first time again after a `git pull` might actually take a while.

I wonder if something similar to `git-lfs` would make sense for standalone installations of HotCRP.

Anyway, thank you so much for the very fast answers, for the help and for the remote debugging.

We're trying to migrate to HotCRP for PhD qualifier paper reviews, which prompted my initial `pull`s -- I'm sure it will be an interesting use case.
